slackbot/dispatcher.py

```diff
@@ -135,13 +135,43 @@ def filter_text(self, msg):
         return msg
 
     def loop(self):
+        # once/second, check events
+        # run idle handlers whenever idle
         while True:
             events = self._client.rtm_read()
             for event in events:
                 if event.get('type') != 'message':
                     continue
                 self._on_new_message(event)
-            time.sleep(1)
+
+            # run idle handlers as long as we've been idle
+            for func in self._plugins.get_idle_plugins():
+                if not func:
+                    continue
+
+                # if actions are pending, don't do anything
+                if not self._pool.queue.empty():
+                    break
+
+                # if some action was taken, don't run the remaining handlers
+                if self._client.idle_time() < 1:
+                    break
+
+                try:
```

Comment:
I wonder if the next few lines should be in loop() or elsewhere. Unlike regular messages, this would mean that idle functions are run in the loop thread. Other messages are handled in their own thread, being added to the pool.